winrm/x509.go

```diff
@@ -0,0 +1,280 @@
+// Copyright 2016 Canonical Ltd.
+// Copyright 2016 Canonical Ltd.
+// Licensed under the LGPLv3, see LICENCE file for details.
+
+package winrm
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/juju/utils"
+	"github.com/juju/utils/cert"
+)
+
+// List of exported internal errors
+var (
+	ErrNoClientCert       = errors.New("No client winrm cert inside dir")
+	ErrNoClientPrivateKey = errors.New("No client private key inside dir")
+	ErrNoX509Folder       = errors.New("No configuration x509 folder is present in this location")
+)
+
+// x509WinRM type that defines windows remote manager
+// credentials for client-server secure communication
+type x509 struct {
+	cert   []byte // client cert
+	key    []byte // client private key
+	cacert []byte // ca server cert
+	err    error  // internal err state
+}
+
+var (
+	mu          sync.Mutex
+	credentials x509 //winrm credentials
```

Comment:
I'm not keen on the use of globals here. I guess this is derived from the ssh/clientkeys.go code, which frankly is a bit ugly. The forces that directed ssh/clientkeys.go do not apply to winrm. In particular, we need keys for ssh *on disk* because we shell out to openssh. We don't do that for winrm.

Can you please change the code to parse and return certs/keys in-memory. I'd also prefer if we separated generation from loading. The caller can intercept an errors.IsNotFound(err), and generate a cert/key pair and write it to disk. I expect we'll only need to do that in one location in Juju, like we do for ssh.LoadClientKeys.